source/server/transport_socket_config_impl.h

```diff
@@ -11,25 +13,38 @@ namespace Configuration {
  */
 class TransportSocketFactoryContextImpl : public TransportSocketFactoryContext {
 public:
-  TransportSocketFactoryContextImpl(Ssl::ContextManager& context_manager, Stats::Scope& stats_scope,
-                                    Secret::SecretManager& secret_manager,
-                                    Init::Manager& init_manager)
-      : context_manager_(context_manager), stats_scope_(stats_scope),
-        secret_manager_(secret_manager), init_manager_(init_manager) {}
+  TransportSocketFactoryContextImpl(
+      Ssl::ContextManager& context_manager, Stats::Scope& stats_scope, Upstream::ClusterManager& cm,
+      Init::Manager& init_manager,
+      Secret::DynamicTlsCertificateSecretProviderFactoryContextPtr secret_provider_context)
+      : context_manager_(context_manager), stats_scope_(stats_scope), cluster_manager_(cm),
+        init_manager_(init_manager), secret_provider_context_(std::move(secret_provider_context)),
+        secret_provider_factory_(*secret_provider_context_, init_manager_) {}
 
   Ssl::ContextManager& sslContextManager() override { return context_manager_; }
 
   Stats::Scope& statsScope() const override { return stats_scope_; }
 
-  Secret::SecretManager& secretManager() override { return secret_manager_; }
-
   Init::Manager& initManager() override { return init_manager_; }
 
+  Upstream::ClusterManager& clusterManager() override { return cluster_manager_; }
+
+  Secret::SecretManager& secretManager() override {
+    return cluster_manager_.clusterManagerFactory().secretManager();
+  }
+
+  Secret::DynamicTlsCertificateSecretProviderFactory&
+  dynamicTlsCertificateSecretProviderFactory() override {
+    return secret_provider_factory_;
+  }
+
 private:
   Ssl::ContextManager& context_manager_;
   Stats::Scope& stats_scope_;
-  Secret::SecretManager& secret_manager_;
+  Upstream::ClusterManager& cluster_manager_;
   Init::Manager& init_manager_;
+  Secret::DynamicTlsCertificateSecretProviderFactoryContextPtr secret_provider_context_;
```

Comment:
not need to hold secret_provider_context_ in the factory_context.   I think the former out-live the latter.